pkg/controller/infrastructure/actuator.go

```diff
@@ -570,6 +572,51 @@ func (a *actuator) Restore(ctx context.Context, log logr.Logger, infra *extensio
 	return a.reconcile(ctx, log, infra, cluster, terraformer.CreateOrUpdateState{State: &terraformState.Data})
 }
 
+func (a *actuator) getDualStackValues(
+	infra *extensionsv1alpha1.Infrastructure,
+	config *alicloudv1alpha1.InfrastructureConfig,
+	credentials *alicloud.Credentials,
+) (*DualStack, error) {
+	dualStack := DualStack{
+		Enabled: false,
+	}
+	if config.DualStack == nil || !config.DualStack.Enabled {
+		return &dualStack, nil
+	}
+	dualStack.Enabled = true
+	dualStack.Zone_A_IPV6_MASK = 255
+	dualStack.Zone_B_IPV6_MASK = 254
+	nlbClient, err := a.newClientFactory.NewNLBClient(infra.Spec.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+	if err != nil {
+		return nil, err
+	}
+	zones, err := nlbClient.GetNLBAvailableZones(infra.Spec.Region)
+	if err != nil {
+		return nil, err
+	}
+	if len(zones) < 2 {
+		return nil, fmt.Errorf("not enough available zones for DualStack")
+	}
+	dualStack.Zone_A = zones[0]
+	dualStack.Zone_B = zones[1]
+
+	// DualStack only for managed vpc
+	// vpcCidr := config.Networks.VPC.CIDR
+
+	subCidr_a, err := getLastSubCidr(*config.Networks.VPC.CIDR, 28, 1)
+	if err != nil {
+		return nil, fmt.Errorf("get sub cidr failed")
+	}
+	subCidr_b, err := getLastSubCidr(*config.Networks.VPC.CIDR, 28, 2)
+	if err != nil {
+		return nil, fmt.Errorf("get sub cidr failed")
+	}
+	dualStack.Zone_A_CIDR = subCidr_a
+	dualStack.Zone_B_CIDR = subCidr_b
+
+	return &dualStack, nil
+}
```

Comment:
In order to enable dual stack, we need to create two additional Vswitche. The main function of the above code is to calculate the necessary parameters for creating Vswitche, including the zone supporting IPV6, the IPV6 MASK, and the IPV4 CIDR, etc.

Reply:
Do we only support two availability zones in alicloud or why is there only ever the mention of zone A and zone B?

What is the IPv6 mask supposed to mean? I doubt that it is meant as a network mask as it would not make a lot of sense to have different size network masks in neighboring zones.

Reply:
Why are the IPv4 ranges always `/28`? 16 addresses is not a lot. What are those ranges used for? It is probably not the amount of nodes as we hopefully can have more than 16 nodes per zone.

Reply:
We are adding two new **vswitch** in preparation for the deployment of Alibaba's NLB. Alibaba requires users to provide two **vswitch** for the deployment of NLB when creating a service that supports IPV6. Therefore, these two **vswitch** need to support IPV6, and obviously, an IPV4 CIDR is also needed. The IPV6 MASK is really confusing, it's the terminology used on the original Alibaba interface, but it actually means subnet identifier. I will rename this variable in the subsequent code. To distinguish from the regular **vswitch**, we choose to use the smallest possible CIDR that is placed at the end for these two additional **vswitch** we created. Since these two **vswitch** are only used for deploying NLB, a /28 CIDR size should be sufficient. However, we could consider expanding it a bit more to facilitate users adding more NLBs.

Reply:
Now IPV6 MASK is renamed with IPV6 Subnet

Reply:
I am still a bit confused. If the `/28` is meant as IPv4 CIDR the 16 addresses are fairly limited. Does this now imply that a cluster can only have at most 16 load balancers? If so, this would be a rather restrictive setting, especially considering that it would need a new extension release to increase the limit.

Could you please provide more details what this range limits? 16 addressed might be sufficient for some things, but might be too low for others.

If the `/28` is meant as an IPv6 CIDR we can skip this discussion altogether.